dom.bs

```diff
@@ -2845,6 +2889,125 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
 </ol>
 
 
+<p><a lt="Other applicable specifications">Specifications</a> may define <dfn export
+id=concept-moving-steps-ext>moving steps</dfn> for all or some <a for=/>nodes</a>. The algorithm is
+passed a <a for=/>node</a> <var ignore>movedNode</var>, and a <a for=/>node</a>-or-null <var
+ignore>oldParent</var> as indicated in the <a for=/>move</a> algorithm below. Like the <span
+data-x="concept-insertion-steps-ext">insertion steps</span>, these steps must not modify the
+<a>node tree</a> that <var>insertedNode</var> <a>participates</a> in, create
+<a for=/>browsing contexts</a>, <a lt="fire an event">fire events</a>, or otherwise execute
+JavaScript. These steps may queue tasks to do these things asynchronously, however.
+
+
+<p>To <dfn export id=concept-node-move>move</dfn> a <var>node</var> into a <var>parent</var> before
+a <var>child</var>, run these steps:
+
+<ol>
+ <!-- Start removing-related bookkeeping steps -->
+ <li><p>Let <var>oldParent</var> be <var>node</var>'s <a for=tree>parent</a>.
+
+ <li><p><a>Assert</a>: <var>oldParent</var> is non-null.
+
+ <li><p>Let <var>index</var> be <var>node</var>'s <a for=tree>index</a>.
+
+ <li>
+  <p>For each <a>live range</a> whose <a for=range>start node</a> is <var>parent</var> and
+  <a for=range>start offset</a> is greater than <var>index</var>, decrease its
+  <a for=range>start offset</a> by 1.</p>
+
+  <p class="note">Note that unlike the traditional <a for=/ lt="remove">removal</a> case, we do not
+  need to update <a>live range</a> state when their <a for=range>start node</a> or
+  <a for=range>end node</a> is an <a>inclusive descendant</a> of the <var>node</var>. This is
+  because said <a>nodes</a> do not get removed from their <a>tree</a>, so ranges associated with
+  them stay intact.</p>
```

Comment:
I don't think so. It should be equivalent to when a range's start container is set (to the first time) to a node after the end container as well. Here are tests for this, which include `intersectsNode` assertions to make sure the Range's state is not peculiar: https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/moveBefore/tentative/live-range-updates.html.

Reply:
Ok, so something else still updates Range boundary points so that start is before end point? The note is then a bit confusing.

Reply:
> Ok, so something else still updates Range boundary points so that start is before end point?

No, a Range's start does not have to be before its end, in DOM tree order. Just like if you create a new Range whose start comes after its end in DOM tree order, the start's [position](https://dom.spec.whatwg.org/#concept-range-bp-position) will be "after" its end, which is observable via `intersectsNode()` since nodes that are after end but before start, will no longer `intersectsNode()` the Range. The same goes for when `moveBefore()` moves things around such that the range's start is after its end, in tree order.

Reply:
Wait, how do you create a Range which has its start after end? setStart/End etc will modify the other boundary point if the order isn't right.

Reply:
Ooof, OK then the spec indeed does leave ranges in a bad state, since what this PR allows (wrt `Range` + `moveBefore()`) is not possible with `Range` today. Great catch, @smaug----. In that case, I think the best way to deal with this would be to, during `moveBefore()`, detect if the moved node is the inclusive ancestor of a start or end boundary point of any range, and if that boundary point crosses its corresponding start/end, and collapse the node at the other boundary point. How does that sound to you?

Reply:
That wouldn't be consistent with other tree operations. I think I'd prefer a model where ranges were treated as if a node was removed and then added back to a different place in tree. 
(Kind of like how MutationRecords are handled)

Reply:
In that case, we wouldn't be able to preserve selection with `moveBefore()`, which is one of the pieces of state we were really hoping to preserve with this API. This is especially annoying when you `moveBefore()` the selection range's start container to somewhere earlier in the doc, or the end container somewhere later. The range is still totally valid, so it seems like we should be able to preserve selection. What do you think?

Reply:
Another option for partial preservation (discussed with @noamr) would be to preserve the selection if and only if the moved node is an inclusive ancestor of _both_ the start and end containers of the range. That way if the selection (or other range) is entirely contained within the moved node, we don't lose it.

Reply:
@smaug---- perhaps a useful middle ground can be that we preserve the selection only if the moved node is an ancestor of the entire range. This way if the range is moved while "intact", the selection would be preserved, but if parts of it are moved it's considered "broken" and it's dissolved. This is useful when moving whole elements which might have bits of text selected inside them.

Reply:
Not sure what  you mean with "wouldn't be able to preserve selection". Using the normal removal/addition rules  would preserve selection the usual way. The range just gets modified.

Reply:
Container approach (not modifying the range if full contained) would be still problematic if the container was moved from light dom to shadow dom, and the range was a range which one can access from selection.getRangeAt().  Selection should not reveal nodes in shadow dom.
Also it might be a bit surprising if selection is preserved in some cases but not in others.

Maybe, unrelated to moveBefore, there should be some helper methods to create a StaticRange from Range, and then update selection later, if/when needed, using that StaticRange. Hmm, though StaticRange isn't quite enough, since one needs also direction.

Reply:
OK the latest commit reverts special range handling, treating ranges just like we would in a normal remove situation. I'll resolve this.